cas-server-support-basic/src/main/java/org/jasig/cas/web/flow/BasicAuthenticationAction.java

```diff
@@ -0,0 +1,92 @@
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(BasicAuthenticationAction.class);
+
+    private static final String AUTHZ_HEADER = "Authorization";
+
+    @Autowired
+    @Qualifier("centralAuthenticationService")
+    private CentralAuthenticationService centralAuthenticationService;
+
+    @Override
+    protected Event doExecute(final RequestContext requestContext) throws Exception {
+
+        final HttpServletRequest request = WebUtils.getHttpServletRequest(requestContext);
+        final HttpServletResponse response = WebUtils.getHttpServletResponse(requestContext);
+
+        final String authHeader = request.getHeader(AUTHZ_HEADER);
+        if (authHeader != null) {
+            final StringTokenizer st = new StringTokenizer(authHeader);
+            if (st.hasMoreTokens()) {
+                final String basic = st.nextToken();
+                if ("basic".equalsIgnoreCase(basic)) {
+                    try {
+                        final String credentials = new String(Base64.decodeBase64(st.nextToken()), "UTF-8");
+                        final int p = credentials.indexOf(':');
+                        if (p != -1) {
+                            final String username = credentials.substring(0, p).trim();
+                            final String password = credentials.substring(p + 1).trim();
+                            final UsernamePasswordCredential credential = new UsernamePasswordCredential(username, password);
+                            LOGGER.debug("Received basic authentication request from credentials {} ", credential);
+                            final TicketGrantingTicket tgt = this.centralAuthenticationService.createTicketGrantingTicket(credential);
+                            WebUtils.putTicketGrantingTicketInScopes(requestContext, tgt);
+                            return success();
+                        } else {
+                            unauthorized(response, "Invalid authentication credentials.");
+                        }
+                    } catch (final UnsupportedEncodingException e) {
+                        throw new Error("Couldn't retrieve authentication", e);
+                    }
+                } else {
+                    LOGGER.warn("{} header does not specify a basic authentication request. Skipping basic authentication challenge",
+                            AUTHZ_HEADER);
+                }
+            } else {
+                LOGGER.warn("{} header is malformed. Skipping basic authentication challenge", AUTHZ_HEADER);
+            }
+        } else {
+            LOGGER.info("No {} header specified. Sending unauthorized challenge for credentials", AUTHZ_HEADER);
+            unauthorized(response);
+        }
+        return error();
+    }
+
+    private void unauthorized(final HttpServletResponse response, final String message) throws IOException {
+        response.setHeader("WWW-Authenticate", "Basic realm=\"CAS\"");
```

Comment:
Only if you access this via the browser, yes it is blocking. This is mostly targeted at command-line clients such as curl where credentials are auto-sent and processed. We still need to do more here, as I realize something like curl may not be able to handle the 302 redirect correctly. So I'll need to investigate this a bit more.